### PR TITLE
Fixes #33577 - Only enable Foreman Puppet plugin with Foreman

### PR DESCRIPTION
**Pull request by ekohl**
If the user passed in `--no-enable-foreman` then the Foreman plugin should not be enabled.

### CODE REVIEW

**Comment by ekohl** [comment]
I chose to change the existing migration so we don't end up in a bad state, but I wonder if we should instead create another migration as well that disables it if foreman itself is disabled. Users who are affected couldn't have upgraded to 3.0.0 (https://community.theforeman.org/t/foreman-proxy-3-0-installation-fails-with-parameter-config-file-group-expects-a-string-value/25438) so perhaps it's ok, but it may not be the cleanest.

**Review by wbclark — COMMENTED** [review]
I would personally vote for a new migration that disables the plugin if foreman itself is disabled, for the reason that it would reduce the number of accidentally broken configurations in the wild (vs. the solution here which stops adding to them but doesn't correct those which already exist)

Perhaps the difference is minimal or low impact but I like the practice of trying to reduce that drift wherever possible, as that can prevent future problems.

**Comment by ekohl** [comment]
Updated. Now also contains the CLI part of it (as reported by the user) and tests. I've started a new pattern for the tests and want to look at using for all other tests. The reporting of what failed is much clearer to me. We can also pull the `let` blocks to a higher level to share them between tests.

File: spec/migration_spec.rb [hunk 1]
```diff
@@ -142,4 +142,38 @@
...
+          expect(migrated_answers['foreman::cli::puppet']).to be false
+        end
+
+        it 'keeps foreman disabled' do
+          expect(migrated_answers['foreman']).to be false
+        end
+
+        it 'adds foreman::plugin::puppet disabled' do
+          expect(migrated_answers['foreman::plugin::puppet']).to be false
+        end
```
Comment (wbclark) [inline spec/migration_spec.rb:175]: :+1: :+1: :+1: 

Fully agreed about using this approach to make test output more readable.

File: config/katello.migrations/170331152302-add-cli.rb [hunk 2]
```diff
@@ -1 +1,3 @@
-answers['foreman::cli'] ||= true
+unless answers.key?('foreman::cli')
+  answers['foreman::cli'] = true
+end
```
Comment (wbclark) [inline config/katello.migrations/170331152302-add-cli.rb:3]: I am guessing there is some subtle difference here. Can you elaborate on the change for this older migration?

Comment (ekohl) [reply config/katello.migrations/170331152302-add-cli.rb:3]: What `||=` does is that if the value evaluates `false` (so `false` or `nil`), it sets the value to `true`.

So previously this happened:
```ruby
answers['foreman::cli'] = false # from the answers
answers['foreman::cli'] ||= true # from the migration
```

So the result was that it was always enabled while the intended migration was to add it with a default value of `true`.

In https://github.com/theforeman/foreman-installer/pull/700 I played around with implementing a hypothetical DSL for migrations that would convey the intention of the migration much better.

Comment (wbclark) [reply config/katello.migrations/170331152302-add-cli.rb:3]: Ah, I understood the behavior of `||=` with `nil` but not `false`. Now I see why this was broken. TY

**Review by wbclark — APPROVED** [review]

**Pull request merged**